x-pack/plugins/security/server/routes/users/change_password.ts

```diff
@@ -27,54 +27,61 @@ export function defineChangeUserPasswordRoutes({
-      const { password, newPassword } = request.body;
-      const isCurrentUser = username === (await authc.getCurrentUser(request))!.username;
+      const { username } = request.params;
+      const { password: currentPassword, newPassword } = request.body;
 
-      // We should prefer `token` over `basic` if possible.
-      const providerToLoginWith = config.authc.providers.includes('token') ? 'token' : 'basic';
+      const currentUser = authc.getCurrentUser(request);
+      const isUserChangingOwnPassword =
+        currentUser && currentUser.username === username && canUserChangePassword(currentUser);
 
-      // If user tries to change own password, let's check if old password is valid first by trying
-      // to login.
-      if (isCurrentUser) {
-        try {
-          const authenticationResult = await authc.login(request, {
-            provider: providerToLoginWith,
-            value: { username, password },
-            // We shouldn't alter authentication state just yet.
-            stateless: true,
-          });
-
-          if (!authenticationResult.succeeded()) {
-            return response.forbidden({ body: authenticationResult.error });
-          }
-        } catch (error) {
-          return response.customError(wrapIntoCustomErrorResponse(error));
-        }
-      }
+      // If user is changing their own password they should provide a proof of knowledge their
+      // current password via sending it in `Authorization: Basic base64(username:current password)`
+      // HTTP header no matter how they logged in to Kibana.
+      const scopedClusterClient = clusterClient.asScoped(
+        isUserChangingOwnPassword
+          ? {
+              headers: {
+                ...request.headers,
+                authorization: `Basic ${Buffer.from(`${username}:${currentPassword}`).toString(
+                  'base64'
+                )}`,
+              },
+            }
+          : request
+      );
 
       try {
-        await clusterClient.asScoped(request).callAsCurrentUser('shield.changePassword', {
+        await scopedClusterClient.callAsCurrentUser('shield.changePassword', {
           username,
           body: { password: newPassword },
         });
+      } catch (error) {
+        // This may happen only if user's credentials are rejected meaning that current password
+        // isn't correct.
+        if (isUserChangingOwnPassword && getErrorStatusCode(error) === 401) {
+          return response.forbidden({ body: error });
+        }
+
+        return response.customError(wrapIntoCustomErrorResponse(error));
+      }
 
-        // Now we authenticate user with the new password again updating current session if any.
-        if (isCurrentUser) {
+      // Now we authenticate user with the new password again updating current session if any.
+      if (isUserChangingOwnPassword) {
+        try {
           const authenticationResult = await authc.login(request, {
```

Comment:
__note to myself:__ we shouldn't re-login if user didn't a session previously (e.g. authenticated via headers).